File: web/src/api/microstructure.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import client from './client';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface LhbRecord {
  id: number;
  trade_date: string;
  ts_code: string;
  name: string;
  close?: number | null;
  pct_change?: number | null;
  turnover_rate?: number | null;
  amount?: number | null;
  lhb_buy_amount?: number | null;
  lhb_sell_amount?: number | null;
  lhb_net_amount?: number | null;
  total_buy?: number | null;
  total_sell?: number | null;
  total_net?: number | null;
  net_buy_amt?: number | null;
  buy_seat_count?: number | null;
  sell_seat_count?: number | null;
  reason: string;
  source: string;
  created_at?: string | null;
}

export interface LhbListResponse {
  items: LhbRecord[];
  total: number;
  page: number;
  page_size: number;
}

export interface HsgtFlow {
  id: number;
  trade_date: string;
  type: string;
  buy_amount?: number | null;
  sell_amount?: number | null;
  net_amount?: number | null;
  balance?: number | null;
  source: string;
  created_at?: string | null;
}

export interface HsgtListResponse {
  items: HsgtFlow[];
  total: number;
}

export interface MarginBalance {
  id: number;
  trade_date: string;
  ts_code: string;
  name: string;
  financing_balance?: number | null;
  financing_buy?: number | null;
  securities_balance?: number | null;
  securities_sell?: number | null;
  exchange: string;
  source: string;
  created_at?: string | null;
}

export interface MarginListResponse {
  items: MarginBalance[];
  total: number;
  page: number;
  page_size: number;
}

export interface RestrictedRelease {
  id: number;
  ts_code: string;
  name: string;
  restricted_date: string;
  restricted_type: string;
  restricted_number?: number | null;
  restricted_amount?: number | null;
  lift_ratio?: number | null;
  source: string;
  created_at?: string | null;
}

export interface RestrictedListResponse {
  items: RestrictedRelease[];
  total: number;
  page: number;
  page_size: number;
}

export interface MicrostructureSummary {
  as_of?: string | null;
  lhb: {
    trade_date?: string | null;
    count?: number;
    top_buyers?: LhbRecord[];
    top_sellers?: LhbRecord[];
  };
  hsgt: {
    trade_date?: string | null;
    north_net?: number | null;
    sh_net?: number | null;
    sz_net?: number | null;
    rows?: HsgtFlow[];
  };
  margin: {
    trade_date?: string | null;
    total_financing_balance?: number;
    total_securities_balance?: number;
  };
  release: {
    upcoming_30d_count?: number;
    upcoming_30d_amount?: number;
  };
}

// ---------------------------------------------------------------------------
// API + hooks
// ---------------------------------------------------------------------------

function buildQueryString(params: Record<string, unknown> | undefined): string {
  if (!params) return '';
  const parts: string[] = [];
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined || value === null || value === '') continue;
    parts.push(`${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`);
  }
  return parts.length === 0 ? '' : `?${parts.join('&')}`;
}

export const microstructureApi = {
  listLhb: (params: Record<string, unknown> = {}) =>
    client.get<LhbListResponse>(`/microstructure/lhb${buildQueryString(params)}`),
  listHsgt: (params: Record<string, unknown> = {}) =>
    client.get<HsgtListResponse>(`/microstructure/hsgt${buildQueryString(params)}`),
  listMargin: (params: Record<string, unknown> = {}) =>
    client.get<MarginListResponse>(`/microstructure/margin${buildQueryString(params)}`),
  listReleases: (params: Record<string, unknown> = {}) =>
    client.get<RestrictedListResponse>(
      `/microstructure/restricted-releases${buildQueryString(params)}`,
    ),
  summary: () => client.get<MicrostructureSummary>('/microstructure/summary'),
  facets: () => client.get<{ exchanges: string[] }>('/microstructure/facets'),
  refresh: () =>
    client.post<{ status: string; records: string; warnings: string[] }>(
      '/microstructure/refresh',
    ),
};

export function useMicrostructureLhb(params: Record<string, unknown> = {}) {
  return useQuery({
    queryKey: ['microstructure', 'lhb', params],
    queryFn: async () => (await microstructureApi.listLhb(params)).data,
    staleTime: 60_000,
  });
}

export function useMicrostructureHsgt(params: Record<string, unknown> = {}) {
  return useQuery({
    queryKey: ['microstructure', 'hsgt', params],
    queryFn: async () => (await microstructureApi.listHsgt(params)).data,
    staleTime: 60_000,
  });
}

export function useMicrostructureMargin(params: Record<string, unknown> = {}) {
  return useQuery({
    queryKey: ['microstructure', 'margin', params],
    queryFn: async () => (await microstructureApi.listMargin(params)).data,
    staleTime: 60_000,
  });
}

export function useMicrostructureReleases(params: Record<string, unknown> = {}) {
  return useQuery({
    queryKey: ['microstructure', 'releases', params],
    queryFn: async () => (await microstructureApi.listReleases(params)).data,
    staleTime: 120_000,
  });
}

export function useMicrostructureSummary() {
  return useQuery({
    queryKey: ['microstructure', 'summary'],
    queryFn: async () => (await microstructureApi.summary()).data,
    staleTime: 60_000,
  });
}

export function useRefreshMicrostructure() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async () => (await microstructureApi.refresh()).data,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['microstructure'] });
    },
  });
}

File: app/schemas/microstructure.py
"""A-share micro-structure data Pydantic schemas.

Used by the API layer to serialize ORM rows for the four micro-structure
data classes (龙虎榜 / 沪深港通 / 融资融券 / 限售解禁).
"""

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# 龙虎榜 (LHB)
# ---------------------------------------------------------------------------


class LhbRecordBase(BaseModel):
    """Common LHB fields exposed in list / detail responses."""

    trade_date: date
    ts_code: str
    name: str
    close: float | None = None
    pct_change: float | None = None
    turnover_rate: float | None = None
    amount: float | None = None

    lhb_buy_amount: float | None = None
    lhb_sell_amount: float | None = None
    lhb_net_amount: float | None = None

    total_buy: float | None = None
    total_sell: float | None = None
    total_net: float | None = None
    net_buy_amt: float | None = None

    buy_seat_count: int | None = None
    sell_seat_count: int | None = None
    reason: str


class LhbRecordOut(LhbRecordBase):
    """LHB record as returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    source: str
    created_at: datetime | None = None


class LhbRecordListResponse(BaseModel):
    """Paginated LHB list response."""

    items: list[LhbRecordOut]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# 沪深港通 (HSGT)
# ---------------------------------------------------------------------------


class HsgtFlowBase(BaseModel):
    """Common HSGT flow fields."""

    trade_date: date
    type: str = Field(..., description="资金流向类型: 北向 / 沪股通 / 深股通 / 南向")
    buy_amount: float | None = None
    sell_amount: float | None = None
    net_amount: float | None = None
    balance: float | None = None


class HsgtFlowOut(HsgtFlowBase):
    """HSGT flow record as returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    source: str
    created_at: datetime | None = None


class HsgtFlowListResponse(BaseModel):
    """HSGT flow list response (no pagination — short window)."""

    items: list[HsgtFlowOut]
    total: int


# ---------------------------------------------------------------------------
# 融资融券 (Margin)
# ---------------------------------------------------------------------------


class MarginBalanceBase(BaseModel):
    """Common margin balance fields."""

    trade_date: date
    ts_code: str
    name: str
    financing_balance: float | None = None
    financing_buy: float | None = None
    securities_balance: float | None = None
    securities_sell: float | None = None
    exchange: str = Field(..., description="交易所: SSE / SZSE")


class MarginBalanceOut(MarginBalanceBase):
    """Margin balance record as returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    source: str
    created_at: datetime | None = None


class MarginBalanceListResponse(BaseModel):
    """Paginated margin balance list response."""

    items: list[MarginBalanceOut]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# 限售解禁 (Restricted Release)
# ---------------------------------------------------------------------------


class RestrictedReleaseBase(BaseModel):
    """Common restricted-release fields."""

    ts_code: str
    name: str
    restricted_date: date
    restricted_type: str = ""
    restricted_number: float | None = None
    restricted_amount: float | None = None
    lift_ratio: float | None = None


class RestrictedReleaseOut(RestrictedReleaseBase):
    """Restricted release record as returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    source: str
    created_at: datetime | None = None


class RestrictedReleaseListResponse(BaseModel):
    """Paginated restricted-release list response."""

    items: list[RestrictedReleaseOut]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# Summary (首页用)
# ---------------------------------------------------------------------------


class MicrostructureSummaryResponse(BaseModel):
    """Latest-day micro-structure summary for the dashboard.

    Each section is best-effort: a missing section means no fresh data
    was available for that data class on the latest trade date.
    """

    as_of: date | None = Field(
        None, description="汇总基准日 (四个分类中最近一个有数据的日期)"
    )
    lhb: dict[str, Any] = Field(
        default_factory=dict,
        description="龙虎榜: { trade_date, count, top_buyers, top_sellers }",
    )
    hsgt: dict[str, Any] = Field(
        default_factory=dict,
        description="北上资金: { trade_date, north_net, sh_net, sz_net }",
    )
    margin: dict[str, Any] = Field(
        default_factory=dict,
        description="融资融券: { trade_date, total_financing_balance, total_securities_balance }",
    )
    release: dict[str, Any] = Field(
        default_factory=dict,
        description="限售解禁: { upcoming_30d_count, upcoming_30d_amount }",
    )
